src/utils/colors.ts

```diff
@@ -1,9 +1,101 @@
-export const getRGBColor = (hex: string, type: string) => {
-  const color = hex.replace(/#/g, '')
-  // rgb values
-  const r = parseInt(color.substr(0, 2), 16)
-  const g = parseInt(color.substr(2, 2), 16)
-  const b = parseInt(color.substr(4, 2), 16)
+import { HspColor, OkclhColor, PrimaryColor } from '@/types'
 
-  return `--tido-color-${type}: ${r}, ${g}, ${b}; --tido-color-${type}-foreground: oklch(0.985 0 0);`
+function hexToOklch(hex: string) {
+  // --- HEX → sRGB ---
+  hex = hex.replace('#', '')
+  if (hex.length === 3) {
+    hex = hex.split('').map(c => c + c).join('')
+  }
+
+  let r = parseInt(hex.slice(0, 2), 16) / 255
+  let g = parseInt(hex.slice(2, 4), 16) / 255
+  let b = parseInt(hex.slice(4, 6), 16) / 255
+
+  // --- sRGB → linear RGB ---
+  const toLinear = c =>
+    c <= 0.04045 ? c / 12.92 : Math.pow((c + 0.055) / 1.055, 2.4)
+
+  r = toLinear(r)
+  g = toLinear(g)
+  b = toLinear(b)
+
+  return linearRgbToOklch(r, g, b)
+}
+
+function hspToOklch(H: number, S: number, P: number) {
+  let R, G, B
+
+  if (S === 0) {
+    R = G = B = P
+  } else {
+    const h = ((H % 360) + 360) % 360 / 60
+    const i = Math.floor(h)
+    const f = h - i
+
+    const p = P * Math.sqrt(1 - S)
+    const q = P * Math.sqrt(1 - S * f)
+    const t = P * Math.sqrt(1 - S * (1 - f))
+
+    switch (i) {
+    case 0: R = P; G = t; B = p; break
+    case 1: R = q; G = P; B = p; break
+    case 2: R = p; G = P; B = t; break
+    case 3: R = p; G = q; B = P; break
+    case 4: R = t; G = p; B = P; break
+    default: R = P; G = p; B = q; break
+    }
+  }
+
+  return linearRgbToOklch(R, G, B)
+}
+
+function linearRgbToOklch(R, G, B) {
+  // --- linear RGB → LMS ---
+  const l = 0.4122214708 * R + 0.5363325363 * G + 0.0514459929 * B
+  const m = 0.2119034982 * R + 0.6806995451 * G + 0.1073969566 * B
+  const s = 0.0883024619 * R + 0.2817188376 * G + 0.6299787005 * B
+
+  const l_ = Math.cbrt(l)
+  const m_ = Math.cbrt(m)
+  const s_ = Math.cbrt(s)
+
+  // --- LMS → OKLab ---
+  const L = 0.2104542553 * l_ + 0.7936177850 * m_ - 0.0040720468 * s_
+  const a = 1.9779984951 * l_ - 2.4285922050 * m_ + 0.4505937099 * s_
+  const b = 0.0259040371 * l_ + 0.7827717662 * m_ - 0.8086757660 * s_
+
+  // --- OKLab → OKLCH ---
+  const C = Math.sqrt(a * a + b * b)
+  let h = Math.atan2(b, a) * 180 / Math.PI
+  if (h < 0) h += 360
+
+  return { l:L, c:C, h }
 }
+
+
+const getAppPrimaryAndForegroundColor = (primaryColor: PrimaryColor, type: string) => {
+  const okclh = getOkclhColor(primaryColor)
+  return `--tido-color-${type}: oklch(${okclh?.l} ${okclh?.c} ${okclh?.h} / 1); --tido-color-${type}-foreground: oklch(0.985 0 0);
+  --annotation-hover: oklch(0.93 0.02 ${okclh?.h}); --annotation-selected: oklch(0.7 0.07 ${okclh?.h})`
```

Comment:
The hover is too light, probably make it a bit darker. Works nice apart from that!